src/reducetensor.cpp

```diff
@@ -427,222 +567,468 @@ void ReduceTensorDescriptor::ReduceTensor(const Handle& handle,
         {
-            param += std::to_string(invariantStrides[i]);
-            if(i < invariantLengths.size() - 1)
+            param += std::to_string(inDescStrides[i]);
+            if(i < inDescStrides.size() - 1)
                 param += ",";
         };
-    }
-    else
-    {
-        param += " -DCK_PARAM_DST_DESC_LENGTHS=1";
-        param += " -DCK_PARAM_DST_DESC_STRIDES=1";
-    };
 
-    param += " -DCK_PARAM_TOREDUCE_DIMS=";
-    for(int i = 0; i < toReduceDims.size(); i++)
-    {
-        param += std::to_string(toReduceDims[i]);
-        if(i < toReduceDims.size() - 1)
-            param += ",";
-    };
+        if(!reduceAllDims)
+        {
+            param += " -DCK_PARAM_DST_DESC_LENGTHS=";
+            for(int i = 0; i < invariantLengths.size(); i++)
+            {
+                param += std::to_string(invariantLengths[i]);
+                if(i < invariantLengths.size() - 1)
+                    param += ",";
+            };
 
-    if(!reduceAllDims)
-    {
-        param += " -DCK_PARAM_INVARIANT_DIMS=";
-        for(int i = 0; i < invariantDims.size(); i++)
+            param += " -DCK_PARAM_DST_DESC_STRIDES=";
+            for(int i = 0; i < invariantStrides.size(); i++)
+            {
+                param += std::to_string(invariantStrides[i]);
+                if(i < invariantLengths.size() - 1)
+                    param += ",";
+            };
+        }
+        else
         {
-            param += std::to_string(invariantDims[i]);
-            if(i < invariantDims.size() - 1)
+            param += " -DCK_PARAM_DST_DESC_LENGTHS=1";
+            param += " -DCK_PARAM_DST_DESC_STRIDES=1";
+        };
+
+        param += " -DCK_PARAM_TOREDUCE_DIMS=";
+        for(int i = 0; i < toReduceDims.size(); i++)
+        {
+            param += std::to_string(toReduceDims[i]);
+            if(i < toReduceDims.size() - 1)
                 param += ",";
         };
-    }
-    else
-        param += " -DCK_PARAM_INVARIANT_DIMS= ";
 
-    param += " -DCK_PARAM_REDUCE_OP=" + std::to_string(detail::GetReduceTensorOpId(reduceOp));
-    param +=
-        " -DCK_PARAM_NAN_PROPAGATE=" + std::to_string(nanPropaOpt == MIOPEN_PROPAGATE_NAN ? 1 : 0);
-    param += " -DCK_PARAM_REDUCE_INDICES=" +
-             std::to_string(reduceIndicesOpt == MIOPEN_REDUCE_TENSOR_FLATTENED_INDICES ? 1 : 0);
+        if(!reduceAllDims)
+        {
+            param += " -DCK_PARAM_INVARIANT_DIMS=";
+            for(int i = 0; i < invariantDims.size(); i++)
+            {
+                param += std::to_string(invariantDims[i]);
+                if(i < invariantDims.size() - 1)
+                    param += ",";
+            };
+        }
+        else
+            param += " -DCK_PARAM_INVARIANT_DIMS= ";
+
+        param += " -DCK_PARAM_REDUCE_OP=" + std::to_string(detail::GetReduceTensorOpId(reduceOp));
+        param += " -DCK_PARAM_NAN_PROPAGATE=" +
+                 std::to_string(nanPropaOpt == MIOPEN_PROPAGATE_NAN ? 1 : 0);
```

Comment:
> [Recommendation] You can avoid conversion of integers to strings:
> 
> ```
> param += " -DCK_PARAM_NAN_PROPAGATE=" + (nanPropaOpt == MIOPEN_PROPAGATE_NAN ? "1" : "0");
> ```

Yes, good

Reply:
> > [Recommendation] You can avoid conversion of integers to strings:
> > ```
> > param += " -DCK_PARAM_NAN_PROPAGATE=" + (nanPropaOpt == MIOPEN_PROPAGATE_NAN ? "1" : "0");
> > ```
> 
> Yes, good

I found this will cause compiler issue, and I can only do so by splitting to two lines as
`param += " -DCK_PARAM_NAN_PROPAGATE=";`
`param += (nanPropaOpt == MIOPEN_PROPAGATE_NAN ? "1" : "0");`

Reply:
Resolved.

[Performance][Quality] @DrizztDoUrden @junliume @qianfengz String manipulations increase technical debt and affect host-side performance a bit. From now on we must block new Solvers that use string manipulations instead of `KernelBuildParameters`.